// x402-solana/dist/utils/index.mjs
// src/utils/helpers.ts
function createPaymentHeaderFromTransaction(transaction, paymentRequirements, x402Version) {
  const serializedTransaction = Buffer.from(transaction.serialize()).toString("base64");
  const paymentPayload = {
    x402Version,
    scheme: paymentRequirements.scheme,
    network: paymentRequirements.network,
    payload: {
      transaction: serializedTransaction
    }
  };
  const paymentHeader = Buffer.from(JSON.stringify(paymentPayload)).toString("base64");
  return paymentHeader;
}
function getDefaultRpcUrl(network) {
  if (network === "solana") {
    return "https://api.mainnet-beta.solana.com";
  } else if (network === "solana-devnet") {
    return "https://api.devnet.solana.com";
  }
  throw new Error(`Unexpected network: ${network}`);
}
function getDefaultTokenAsset(network) {
  if (network === "solana") {
    return {
      address: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
      decimals: 6
    };
  } else if (network === "solana-devnet") {
    return {
      address: "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU",
      decimals: 6
    };
  }
  throw new Error(`Unexpected network: ${network}`);
}
function toAtomicUnits(amount, decimals) {
  return BigInt(Math.floor(amount * Math.pow(10, decimals)));
}
function fromAtomicUnits(atomicUnits, decimals) {
  return Number(atomicUnits) / Math.pow(10, decimals);
}
function usdToMicroUsdc(usdAmount) {
  return Math.floor(usdAmount * 1e6);
}
function microUsdcToUsd(microUsdc) {
  return microUsdc / 1e6;
}

export { createPaymentHeaderFromTransaction, fromAtomicUnits, getDefaultRpcUrl, getDefaultTokenAsset, microUsdcToUsd, toAtomicUnits, usdToMicroUsdc };
