// x402-solana/dist/types/index.js
'use strict';

var types = require('x402/types');
var zod = require('zod');

// src/types/index.ts
var SolanaNetworkSchema = zod.z.enum([
  "solana-devnet",
  "solana"
]);
var SolanaPaymentPayloadSchema = zod.z.object({
  x402Version: zod.z.literal(1),
  scheme: zod.z.literal("exact"),
  network: SolanaNetworkSchema,
  payload: types.ExactSvmPayloadSchema
  // Official SVM payload from x402
});

// src/types/solana-payment.ts
function isSolanaNetwork(network) {
  return network === "solana" || network === "solana-devnet";
}

Object.defineProperty(exports, "ExactSvmPayloadSchema", {
  enumerable: true,
  get: function () { return types.ExactSvmPayloadSchema; }
});
Object.defineProperty(exports, "PaymentRequirementsSchema", {
  enumerable: true,
  get: function () { return types.PaymentRequirementsSchema; }
});
Object.defineProperty(exports, "SettleResponseSchema", {
  enumerable: true,
  get: function () { return types.SettleResponseSchema; }
});
Object.defineProperty(exports, "SupportedPaymentKindSchema", {
  enumerable: true,
  get: function () { return types.SupportedPaymentKindSchema; }
});
Object.defineProperty(exports, "SupportedPaymentKindsResponseSchema", {
  enumerable: true,
  get: function () { return types.SupportedPaymentKindsResponseSchema; }
});
Object.defineProperty(exports, "SupportedSVMNetworks", {
  enumerable: true,
  get: function () { return types.SupportedSVMNetworks; }
});
Object.defineProperty(exports, "SvmNetworkToChainId", {
  enumerable: true,
  get: function () { return types.SvmNetworkToChainId; }
});
Object.defineProperty(exports, "VerifyResponseSchema", {
  enumerable: true,
  get: function () { return types.VerifyResponseSchema; }
});
Object.defineProperty(exports, "x402ResponseSchema", {
  enumerable: true,
  get: function () { return types.x402ResponseSchema; }
});
exports.SolanaNetworkSchema = SolanaNetworkSchema;
exports.SolanaPaymentPayloadSchema = SolanaPaymentPayloadSchema;
exports.isSolanaNetwork = isSolanaNetwork;
